Add KitchenSink.Algorithms.binary_search & tests

mikestok opened the pull request:
The ability to use a "special" calculation for mid point is for @maedhr .

## Changes since 1.0.2

  * Added CHANGELOG.md
  * Added KitchenSink.Algorithms
    * binary_search/{2,3,4}
  * Updated credo 0.6.1 → 0.8.1
  * Updated dialyxir 0.4.4 → 0.5.0
  * Updated ex_doc 0.14.5 → 0.16.1

mikestok committed:
Add KitchenSink.Algorithms.binary_search & tests

(Mike & Marius) and Craig were working separately on implementations of
a binary search for use in the app. This is an attempt to unify those
efforts.

mikestok committed:
Update dev dependencies, add CHANGELOG.md, bump to 1.1.0

Minor version bump as we're adding something to the interface.

(see CHANGELOG.md ;-)

olafura reviewed (changes requested):
Rename the function to match behaviour. Also have the lookup function with two params ;)

mikestok committed:
Rename the interval version of binary_search (PR change)

As the binary search and interval binary search have very different
behaviours which are triggered by the type of the first argument, the
interval one has been renamed to `binary_search_interval` to avoid
"accidental" misuse.

`binary_search_interval` example is more in line with the real use case
for @kc.

Make binary_interval_search and binary_search arguments more consistent

  start, finish, fit, target, ...

Fit functions now take 2 args, the current position and the target
value.

mikestok commented on lib/kitchen_sink/algorithms.ex:
```diff
@@ -0,0 +1,114 @@
+  @spec binary_search(Range.t, fit_func, mid_func) :: binary_search_result
+  def binary_search(start .. finish, fit) do
+    binary_search(start .. finish, fit, &binary_search_midpoint/2)
+  end
```
Is there a way to default that last argument "nicely"?

olafura replied:
`def binary_search(start .. finish, fit \\ &binary_search_midpoint/2)`
